### ISSUE TITLE
XDC loss funcation and Clustering algorithm code 

### ISSUE THREAD

**Issue by say2sarwar** [HumamAlwassel/XDC#8]
I am trying to implement this paper but i am getting some issues in XDC loss function and clustering algorithm. 

Can you please provide the code for loss function.

**Comment by HumamAlwassel** [comment]
Hello @say2sarwar,

Thanks for your interest in our work.

The loss function we use is the cross-entropy loss (implemented in all deep learning frameworks) and the clustering method we use is the k-means algorithm (also implemented in standard Python packages such as Scipy). 

Hope this helps!

**Comment by say2sarwar** [comment]
Batch size is 32 in most cases but the number of clusters is 128, 256 etc. How did you manage to use high cluster size with small batch size? 
Did you produce feature for multiple batches and use kmeans? or any other method?

**Comment by HumamAlwassel** [comment]
The clustering is done offline on precomputed features. 
In each XDC iteration,  
1. We precompute the clip features and re-culster offline to generate pseudo labels
2. We train on the pseudo labels until *the validation loss saturates*
3. Repeat

**Comment by say2sarwar** [comment]
Thank you so much for detail reply.

Let me get this.

Compute features for whole data and use kmeans to produce pseudo labels
use that labels in XDC iteration.

**Comment by HumamAlwassel** [comment]
correct
